queue/driver_mongo.go

```diff
@@ -594,37 +594,26 @@ func isMongoDupKey(err error) bool {
 }
 
 func (d *mongoDriver) Save(ctx context.Context, j amboy.Job) error {
-	job, err := d.prepareInterchange(j)
+	ji, err := d.prepareInterchange(j)
 	if err != nil {
 		return errors.WithStack(err)
 	}
 
-	job.Scopes = j.Scopes()
+	ji.Scopes = j.Scopes()
 
-	return errors.WithStack(d.doUpdate(ctx, job))
+	return errors.WithStack(d.doUpdate(ctx, ji))
 }
 
-func (d *mongoDriver) SaveAndPut(ctx context.Context, toSave amboy.Job, toPut amboy.Job) error {
+func (d *mongoDriver) CompleteAndPut(ctx context.Context, toComplete amboy.Job, toPut amboy.Job) error {
 	sess, err := d.client.StartSession()
 	if err != nil {
 		return errors.Wrap(err, "starting transaction session")
 	}
 	defer sess.EndSession(ctx)
 
-	atomicSaveAndPut := func(sessCtx mongo.SessionContext) (interface{}, error) {
-		// TODO (EVG-13540): for a long-running job that has been lock pinged,
-		// we need to make sure that the modTS/modcount is okay to use. Save()
-		// should still succeed as long as:
-		// * Continuity of in-memory job: the dispatcher shares the same
-		// in-memory copy of the job as the retry handler.
-		// * Concurrency: the dispatcher never runs concurrently with the retry
-		// handler (which would modify modTS/modcount).
-		// * Precondition: the job is complete when this is called.
-		// We should verify these statements are true when doing tests of the
-		// retry handler.
-
-		if err = d.Save(sessCtx, toSave); err != nil {
-			return nil, errors.Wrap(err, "saving old job")
+	atomicCompleteAndPut := func(sessCtx mongo.SessionContext) (interface{}, error) {
+		if err = d.Complete(sessCtx, toComplete); err != nil {
```

Comment:
The TODO is a bit outdated. I don't think this will be a concern because the dispatcher is already done by the time this code is reached and all threads hold onto the same in-memory job from the moment it leaves the queue to the moment the retry handler receives it. Any other concurrency concerns between retry handlers on different hosts will be handled by [EVG-13784](https://jira.mongodb.org/browse/EVG-13784).